dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/WorkContextLoader.java

```diff
@@ -0,0 +1,177 @@
+        this.serviceDelegatorSupplier = serviceDelegatorSupplier;
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Transactional( readOnly = true )
+    public WorkContext load( ImportOptions importOptions, List<Event> events )
+    {
+        sessionFactory.getCurrentSession().flush();
+
+        ImportOptions localImportOptions = importOptions;
+        // API allows a null Import Options
+        if ( localImportOptions == null )
+        {
+            localImportOptions = ImportOptions.getDefaultImportOptions();
+        }
+
+        initializeUser( localImportOptions );
+
+        // Make sure all events have the 'uid' field populated
+        events = uidGen.assignUidToEvents( events );
+
+        final Map<String, ProgramStageInstance> programStageInstanceMap = programStageInstanceSupplier
+            .get( localImportOptions, events );
+
+        final Map<String, Pair<TrackedEntityInstance, Boolean>> teiMap = trackedEntityInstanceSupplier
+            .get( localImportOptions, events );
+
+        return WorkContext.builder()
+        // @formatter:off
+            .importOptions( localImportOptions )
+            .programsMap( programSupplier.get( localImportOptions, events ) )
+            .programStageInstanceMap( programStageInstanceMap )
+            .organisationUnitMap( organisationUnitSupplier.get( localImportOptions, events ) )
+            .trackedEntityInstanceMap( teiMap )
+            .programInstanceMap( programInstanceSupplier.get( localImportOptions, teiMap, events ) )
+            .categoryOptionComboMap( categoryOptionComboSupplier.get( localImportOptions, events ) )
+            .dataElementMap( dataElementSupplier.get( localImportOptions, events ) )
+            .notesMap( noteSupplier.get( localImportOptions, events ) )
+            .assignedUserMap( assignedUserSupplier.get( localImportOptions, events ) )
+            .eventDataValueMap( new EventDataValueAggregator().aggregateDataValues( events, programStageInstanceMap, localImportOptions ))
+            .serviceDelegator( serviceDelegatorSupplier.get() )
+            .build();
+        // @formatter:on
+    }
+
+    /**
+     * Make sure that the {@see User} object's properties are properly initialized,
+     * to avoid running into Hibernate-related issues during validation
+     *
+     * @param importOptions the {@see ImportOptions} object
+     */
+    private void initializeUser( ImportOptions importOptions )
+    {
+        if ( importOptions.getUser() == null )
+        {
+            final User currentUser = this.serviceDelegatorSupplier.get().getCurrentUserService().getCurrentUser();
+
+            //
+            // This should never really happen!
+            //
+            if ( currentUser != null )
+            {
+                UserCredentials userCredentials = currentUser.getUserCredentials();
+                HibernateUtils.initializeProxy( userCredentials );
+                userCredentials.isSuper();
```

Comment:
This is required to avoid an Hibernate error down the line, when we access this method over and over, but since the object is detached, we "pre-load" the data here.

Reply:
Adding comment.